src/middlewared/middlewared/plugins/cloud_backup/snapshot.py

```diff
@@ -0,0 +1,101 @@
+from datetime import datetime
+import json
+import subprocess
+
+from middlewared.plugins.cloud_backup.restic import get_restic_config
+from middlewared.schema import accepts, Bool, Datetime, Dict, Int, List, returns, Str
+from middlewared.service import CallError, Service
+from middlewared.validators import NotMatch
+
+
+class CloudBackupService(Service):
+
+    class Config:
+        cli_namespace = "task.cloud_backup"
+        namespace = "cloud_backup"
+
+    @accepts(Int("id"))
+    @returns(
+        List("cloud_backup_snapshots", items=[
+            Dict(
+                "cloud_backup_snapshot",
+                Str("id"),
+                Str("hostname"),
+                Datetime("time"),
+                List("paths", items=[Str("path")]),
+                additional_attrs=True,
+            ),
+        ]),
+    )
+    def list_snapshots(self, id_):
+        """
+        List existing snapshots for the cloud backup job `id`.
+        """
+        self.middleware.call_sync("network.general.will_perform_activity", "cloud_backup")
+
+        cloud_backup = self.middleware.call_sync("cloud_backup.get_instance", id_)
+
+        restic_config = get_restic_config(cloud_backup)
+
+        try:
+            snapshots = json.loads(subprocess.run(
+                restic_config.cmd + ["--json", "snapshots"],
+                env=restic_config.env,
+                capture_output=True,
+                text=True,
+                check=True,
+            ).stdout)
+        except subprocess.CalledProcessError as e:
+            raise CallError(e.stderr)
+
+        for snapshot in snapshots:
+            snapshot["time"] = datetime.fromisoformat(snapshot["time"])
+
+        return snapshots
+
+    @accepts(Int("id"), Str("snapshot_id", validators=[NotMatch(r"^-")]), Str("path", validators=[NotMatch(r"^-")]))
+    @returns(
+        List("cloud_backup_snapshot_items", items=[
+            Dict(
+                "cloud_backup_snapshot_item",
+                Str("name"),
+                Str("path"),
+                Str("type", enum=["dir", "file"]),
+                Int("size"),
+                Datetime("mtime"),
+                additional_attrs=True,
+            ),
+        ]),
+    )
+    def list_snapshot_directory(self, id_, snapshot_id, path):
```

Comment:
Have we tested what performance is like when dir contains 100K files? 1 million files? Does it need to be a job?

Reply:
It's not a job in the original cloud sync plugin and it never caused any issues. People don't use this to list directories with 100k files (and we can't offer them to do so since no pagination is available).